sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/json/InferSchema.scala

```diff
@@ -246,12 +263,39 @@ private[sql] object InferSchema {
           }
 
         case (StructType(fields1), StructType(fields2)) =>
-          val newFields = (fields1 ++ fields2).groupBy(field => field.name).map {
-            case (name, fieldTypes) =>
-              val dataType = fieldTypes.view.map(_.dataType).reduce(compatibleType)
-              StructField(name, dataType, nullable = true)
+          // Both fields1 and fields2 should be sorted by name, since inferField performs sorting.
+          // Therefore, we can take advantage of the fact that we're merging sorted lists and skip
+          // building a hash map or performing additional sorting.
+          val newFields = new mutable.ArrayBuffer[StructField]()
+
+          var f1Idx = 0
+          var f2Idx = 0
+
+          while (f1Idx < fields1.length && f2Idx < fields2.length) {
+            val f1Name = fields1(f1Idx).name
+            val f2Name = fields2(f2Idx).name
+            if (f1Name == f2Name) {
+              val dataType = compatibleType(fields1(f1Idx).dataType, fields2(f2Idx).dataType)
+              newFields += StructField(f1Name, dataType, nullable = true)
+              f1Idx += 1
+              f2Idx += 1
+            } else if (f1Name < f2Name) {
+              newFields += fields1(f1Idx)
+              f1Idx += 1
+            } else { // f1Name > f2Name
+              newFields += fields2(f2Idx)
+              f2Idx += 1
+            }
+          }
+          while (f1Idx < fields1.length) {
+            newFields += fields1(f1Idx)
+            f1Idx += 1
+          }
+          while (f2Idx < fields2.length) {
+            newFields += fields2(f2Idx)
+            f2Idx += 1
           }
-          StructType(newFields.toSeq.sortBy(_.name))
+          StructType(newFields.toSeq)
```

Comment:
`StructType(Seq[StructField])` just calls `toArray` on the fields... so I would use `toArray` instead of `toSeq` to reduce one allocation.